1-Draft/RFCNNNN-Make-Terminating-Errors-Terminate-The-Right-Way.md

```diff
@@ -0,0 +1,193 @@
+---
+RFC: RFCnnnn
+Author: Kirk Munro
+Status: Draft
+SupercededBy: 
+Version: 1.0
+Area: Engine
+Comments Due: July 15, 2019
+Plan to implement: Yes
+---
+
+# Make terminating errors terminate the right way in PowerShell
+
+By default in PowerShell, terminating errors do not actually terminate. For example, if you invoke this command in global scope, you will see the output "Why?" after the terminating error caused by the previous command:
+
+```PowerShell
+& {
+    $string = 'Hello'
+    $string.Substring(99)
+    'Why?'
+}
+```
+
+PowerShell has upheld this behaviour since version 1.0 of the language. You can make the terminating error actually terminate execution of the command, by wrapping the command in try/catch, like this:
+
+```PowerShell
+try {
+    $string = 'Hello'
+    $string.Substring(99)
+    'Why?'
+} catch {
+    throw
+}
+```
+
+You can also convert the terminating-yet-handled-as-a-non-terminating error into an actual terminating error like this:
+
+```PowerShell
+& {
+    $ErrorActionPreference = 'Stop'
+    $string = 'Hello'
+    $string.Substring(99)
+    'Why?'
+}
+```
+
+In those last two examples, the exception raised by the .NET method terminates execution of the running command.
+
+The difference between the first example and the workarounds poses a risk to scripters who share scripts or modules with the community.
+
+In the first workaround, the risk is that end users using a shared resource such as a script or module may see different behaviour from the logic within that module depending on whether or not they were inside of a `try` block when they invoked the script or a command exported by the module. That risk is very undesirable, and as a result many community members who share scripts/modules with the community wrap their logic in a `try/catch{throw}` (or similar) scaffolding to ensure that the behavior of their code is consistent no matter where or how it was invoked.
+
+In the second workaround, if the shared script does not use `$ErrorActionPreference = 'Stop'`, a caller can get different behaviour by manipulating their `$ErrorActionPreference`. The caller should not be able to manipulate terminating error behavior in commands that they invoke -- that's up to the command author, and they shouldn't have to use extra handling to make terminating errors terminate.
+
+Now consider this code snippet:
+
+```PowerShell
+New-Module -Name ThisShouldNotImport {
+    $myList = [System.Collections.Generics.List[string]]::new()
```

Comment:
>Generics [](start = 34, length = 8)

Would be nice to make it more obvious that this is wrong... `[System.Collections.Typo.List[string]]` or somesuch.

Reply:
How would you like to see that? It can't necessarily be done by the parser, because that code may be dependent on assemblies that are loaded as part of the module loading process.

Are you proposing a fuzzy match algorithm to pick up on typos in various things at runtime if an error occurs where something is not found, such as a type name, variable name, command name, etc.? If so, you should log that as an issue in the PowerShell repository.

Reply:
No; I was just suggesting that the typo in the example could be made more obvious to human readers. Just a cosmetic suggestion to aid people.

Reply:
Ah, that's a great point. I like the subtlety of the issue, so maybe I'll add a comment before it to make it clear. Thanks. :)